compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -760,24 +776,41 @@ fn infer_placeholder_type(
             // The parser provided a sub-optimal `HasPlaceholders` suggestion for the type.
             // We are typeck and have the real type, so remove that and suggest the actual type.
             err.suggestions.clear();
-            err.span_suggestion(
-                span,
-                "provide a type for the item",
-                format!("{}: {}", item_ident, ty),
-                Applicability::MachineApplicable,
-            )
-            .emit_unless(ty.references_error());
+
+            // Suggesting unnameable types won't help.
+            if !contains_anonymous(ty) {
```

Comment:
Don't put a `!` when you have both `if` and `else` because it can just be confusing. Easier to understand if you flip this, then you would think "if type contains anonymous types, then note that is cannot be named, if not, then suggest the correct type" instead of "if it does not (`!`) contain anonymous types, then ..., and if it does, then ..."

Reply:
I disagree. The then-branch is the "happy"/expected path, whereas the else-branch handles the corner case. In my view, it is easier to read if the general case comes first, and the special/corner case afterwards.